piker/brokers/kucoin.py

```diff
@@ -0,0 +1,670 @@
+                # and that history has been written
+                sym: {
+                    'symbol_info': {
+                        'asset_type': 'crypto',
+                        'price_tick_size': 0.0005,
+                        'lot_tick_size': 0.1,
+                    },
+                    'shm_write_opts': {'sum_tick_vml': False},
+                    'fqsn': sym,
+                },
+            }
+            @acm
+            async def subscribe(ws: wsproto.WSConnection):
+                @acm
+                async def open_ping_task(ws: wsproto.WSConnection):
+                    '''
+                    Spawn a non-blocking task that pings the ws
+                    server every ping_interval so Kucoin doesn't drop
+                    our connection
+
+                    '''
+                    async with trio.open_nursery() as n:
+                        # TODO: cache this task so it's only called once
+                        async def ping_server():
+                            while True:
+                                await trio.sleep((ping_interval - 1000) / 1000)
+                                await ws.send_msg({'id': connect_id, 'type': 'ping'})
+
+                        log.info(f'Starting ping task for {sym}')
+                        n.start_soon(ping_server)
+
+                        yield ws
+
+                        n.cancel_scope.cancel()
+
+                # Spawn the ping task here
+                async with open_ping_task(ws) as ws:
+                    tasks = []
+                    tasks.append(make_sub(kucoin_sym, connect_id, level='l3'))
+                    tasks.append(make_sub(kucoin_sym, connect_id, level='l1'))
+
+                    for task in tasks:
+                        log.info(
+                            f'Subscribing to {task["topic"]} feed for {sym}')
+                        await ws.send_msg(task)
+
+                    yield
+
+                    # unsub
+                    if ws.connected():
+                        log.info(f'Unsubscribing to {kucoin_sym} feed')
+                        await ws.send_msg(
+                            {
+                                'id': connect_id,
+                                'type': 'unsubscribe',
+                                'topic': f'/market/ticker:{sym}',
+                                'privateChannel': False,
+                                'response': True,
+                            }
+                        )
+
+            async with (
+                open_autorecon_ws(
+                    f'wss://ws-api-spot.kucoin.com/?token={token}&[connectId={connect_id}]',
+                    fixture=subscribe,
+                ) as ws,
+                stream_messages(ws, sym) as msg_gen,
+            ):
+                typ, quote = await anext(msg_gen)
+
+                while typ != 'trade':
+                    # TODO: use ``anext()`` when it lands in 3.10!
```

Comment:
you can drop this comment now, also maybe put a comment about how we don't unblock until we get a first "trade quote"